Add logStores method and various other improvements

rhgills opened the pull request.

dfed commented:
LGTM. Though let's figure out that build issue.

dfed commented:
Rebased you off of master. You were missing your .travis.yml file. Merge when green

rhgills commented:
Made some more changes in:
https://github.com/square/Aardvark/commit/dca365f436372e0b7d8478d7a65b7d17e171060f
and 
https://github.com/square/Aardvark/commit/d153981e9a92834e158bd79de6d38bdc6819abaa.

In dca365f436372e0b7d8478d7a65b7d17e171060f:
- Adding the default bug reporting gesture was creating a defaultLogStore, instead of using the dispatch_once'd one, which had the new properties. No reason for that, as it set no special configuration. Changed to request the defaultLogStore from the distributor instead.

- Tests were failing because the `defaultLogStore` implementation was calling through to itself. In addition, from the dispatch_once docs:

"The result of using a predicate with automatic or dynamic storage (including Objective-C instance variables) is undefined."

I replaced `dispatch_once` using an ivar with a `BOOL` to determine if this is the first time the accessor is called. If the accessor is called for the first time, then we'll lazily instantiate, as before.

In d153981e9a92834e158bd79de6d38bdc6819abaa:
Made the ARKLogTableViewController only set its title if it doesn't already have one. Want to be able to set more specific titles than "Logs" for when we're using multiple ARKLogTableViewControllers in our app.

Let me know if still approved!

dfed commented:
Love the work with the `dispatch_once`... that was on my todo list to fix. However it needs some thread-safety improvements before it can be merged.

dfed commented on Log Viewing/ARKLogTableViewController.m:
```diff
@@ -109,16 +109,17 @@ - (void)viewWillDisappear:(BOOL)animated;
 
-- (void)viewDidLoad;
-{
+- (void)viewDidLoad; {
```
non-blocking sylte nit: newline after `;`

dfed commented on Log Viewing/ARKLogTableViewController.m:
```diff
@@ -109,16 +109,17 @@ - (void)viewWillDisappear:(BOOL)animated;
-    
-    self.title = NSLocalizedString(@"Logs", @"Title of log viewing screen.");
+
+    if (!self.title) {
```
`self.title.length == 0`. Unless you specifically only want to change this if the title is nil, at which point I'd prefer a `!= nil`

dfed commented on Logging/ARKLogDistributor.h:
```diff
@@ -38,6 +38,9 @@
 @property (atomic, strong, readwrite) ARKLogStore *defaultLogStore;
 
+/// Returns all instances of `ARKLogStore` that are currently registered as observers on this log distributor.
+@property (atomic, strong, readonly) NSSet *logStores;
```
`(copy, readonly)`

dfed commented:
Oh, and please update the version number in the pod spec to 1.0.2

rhgills commented:
All feedback should be addressed. Please take another look!

dfed commented on AardvarkTests/ARKLogDistributorTests.m:
```diff
@@ -143,7 +143,13 @@ - (void)test_setLogMessageClass_appendedLogsAreCorrectClass;
+    XCTAssertFalse(defaultLogStore.prefixNameWhenPrintingToConsole);
+
+    // Should return the same instance on subsequent property accesses.
+    XCTAssertEqual(logDistributor.defaultLogStore, defaultLogStore);
```
nice

dfed commented on Logging/ARKLogDistributor.m:
```diff
@@ -34,9 +34,13 @@ @interface ARKLogDistributor ()
 
 @property (atomic, assign) Class internalLogMessageClass;
 @property (atomic, weak) ARKLogStore *weakDefaultLogStore;
+@property (atomic, strong, readonly) NSRecursiveLock *defaultLogStorePropertyLock;
```
non-blocking style nit: don't need the `strong` (#5 removes all the excessive modifiers)

dfed commented on Logging/ARKLogDistributor.m:
```diff
@@ -80,6 +83,9 @@ - (instancetype)init;
     
     _logObservers = [NSMutableArray new];
+
+    _defaultLogStorePropertyLock = [NSRecursiveLock new];
```
Why a recursive lock? You aren't expecting reentrancy are you?

dfed commented on Logging/ARKLogDistributor.m:
```diff
@@ -91,28 +97,43 @@ - (instancetype)init;
-    if (logStore) {
-        // Add the new log store. The logObserver array will hold onto the log store strongly.
-        [self addLogObserver:logStore];
+    [self.defaultLogStorePropertyLock lock];
```
Technically this lock only needs to be around setting the `weakDefaultLogStore`, but I kinda dig having it around the whole process

rhgills replied:
:+1:

dfed commented on Logging/ARKLogDistributor.m:
```diff
@@ -91,28 +97,43 @@ - (instancetype)init;
+            ARKLogStore *defaultLogStore = [[ARKLogStore alloc] initWithPersistedLogFileName:[NSStringFromClass([self class]) stringByAppendingString:@"_DefaultLogStore"]];
+            defaultLogStore.name = @"Default";
+            defaultLogStore.prefixNameWhenPrintingToConsole = NO;
+            self.defaultLogStore = defaultLogStore;
```
ah, here's the reentrancy

dfed commented:
Looks like there's only the non-blocking style nit. LGTM. Merge when green!

rhgills commented:
merging! thanks!
